tf_quant_finance/math/jacobian.py

```diff
@@ -0,0 +1,129 @@
+# Lint as: python3
+# Copyright 2020 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Helper functions for computing jacobian."""
+
+
+import functools
+
+import tensorflow.compat.v2 as tf
+
+
+def jacobian(func, xs, unconnected_gradients=None, name=None):
+  """Computes the jacobian of `func` wrt to `*xs`.
+
+  Args:
+   func: Python callable accepting one `Tensor` of shape of `x` and returning
+      a `Tensor` of any shape. The function whose jacobian is to be computed.
+    xs: Python list of parameters of `f` for which to differentiate. (Can also
+      be single `Tensor`.)
+    unconnected_gradients: An enum `tf.UnconnectedGradients` which specifies the
+      gradient value returned when the given input tensors are unconnected.
+      Default value: `None`, which maps to `tf.UnconnectedGradients.NONE`.
+    name: Python `str` name prefixed to ops created by this function.
+      Default value: `None` (i.e., 'jacobian').
+
+  Returns:
+    A `Tensor` with the gradient of `y` wrt each of `xs` or a list of `Tensor`s
+    if `xs` is a list.
+  """
+  unconnected_gradients = unconnected_gradients or tf.UnconnectedGradients.NONE
+  with tf.name_scope(name or "jacobian"):
+    xs, is_xs_list_like = _prepare_args(xs)
+    if not callable(func):
+      raise ValueError("`func_or_y` should be a callable in eager mode or "
+                       "when `tf.GradientTape` is used.")
+    with tf.GradientTape() as tape:
+      for x in xs:
+        tape.watch(x)
+      y = func(*xs)
+    jac = tape.jacobian(y, xs,
+                        unconnected_gradients=unconnected_gradients)
```

Comment:
I think we should use [batch_jacobian](https://www.tensorflow.org/api_docs/python/tf/GradientTape#batch_jacobian here. Otherwise, the shapes will explode once you try computing this with a batch shape (see the comment to one of the tests). The first dimension should be treated as a batch dimension.
If we change this to batch_jacobian, please add the extra arguments that come with this function (parallel_iterations, experimental_use_pfor) and set their default values

Reply:
Thanks for your suggestion and guided notebook. Addressed in ec3710e